map GLib.timeout -> g_timeout_add and GLib.timeout_seconds -> g_timeout_add_seconds

3n-k1 opened the pull request:
breaking changes wrt the side effects of GLib.timeout now using milliseconds instead of seconds
GLib.timeout has been renamed to Glib.timeout_seconds
GLib::Priority has been added to map LibGLib::PRIORITY_* constants
GLib.timeout and GLib.timeout_seconds now take a priority argument that defaults to GLib::Priority::Default

i think that the most intuitive thing for GLib.timeout to call is g_timeout_add, with a new GLib.timeout_seconds calling g_timeout_add_seconds. the timeout, timeout_seconds, and idle_add methods have also been changed to take an optional priority argument that defaults to LibGLib::PRIORITY_DEFAULT, which is what [g_timeout_add uses](https://developer.gnome.org/glib/stable/glib-The-Main-Event-Loop.html#g-timeout-add)

3n-k1 commented:
also i'm totally gonna be that person and ask if you could add the `hacktoberfest-accepted` label to this pr :p

jhass commented:
Looking quite good!

I'm thinking, what do you think how evil - in terms of the different guarantees the two methods provide - would it be to instead have one method taking a [`Time::Span`](https://crystal-lang.org/api/0.35.1/Time/Span.html) and check whether its subsecond component is 0, picking `g_timeout_add_seconds` if so and `g_timeout_add`otherwise?

But my biggest concern with your current proposal is that it very subtly breaks existing code, turning previous seconds long timeouts into milliseconds long ones.

3n-k1 commented:
i'm not sure i like the idea of using a `Span` object, but i'm still learning crystal's idioms. i also think that would open up the possibility for some weird bugs if that second heuristic happened, since i've read that `g_timeout_add` and `g_timeout_add_seconds` work slightly differently

i agree that this change is a bit *too* breaking, and i think any changes to `GLib.timeout` will be. perhaps we could add `GLib.timeout_seconds` and `GLib.timeout_milliseconds` methods, and very slowly deprecate `GLib.timeout`?

jhass commented:
I'm fine with breaking immediately as long as it throws a compile time error. The library is in `0.x` state after all :) 

The advantage of a `Time::Span` based API is that you could do thinks like `GLib.timeout 1.minute`, which reads quite nicely I think :)

3n-k1 commented:
you're right, that does read nicely. maybe something like this could work?

```crystal
def self.timeout(priority = ..., time : Time::Span, &block : -> Bool):
  if time.milliseconds
    # if this overflows, let the exception happen
    ms = time.total_milliseconds
    LibGLib.timeout_add(...)
  else
    sec = time.total_seconds
    LibGLib.timeout_add_seconds(...)
  end
end
```

apologies for any typos, I'm on a phone right now

jhass commented:
Sorry, forgot to answer 🤦 

Looks alright to me, I think we do should handle the overflow like I mentioned above if we combine both methods. Also we should document these behaviors :)

3n-k1 commented:
no worries, my memory is finicky too :p

i really think handling overflow like that is a bad idea. g_timeout_add and g_timeout_add_seconds work slightly differently, so i think silently bumping the timeout to add_seconds goes against the principle of least surprise. i think it would be better to either fail on overflow (at compile time if possible) or add some kind of `handle_overflow` flag to `self.timeout` that defaults to false

jhass commented:
Well, by that argument the original idea of combining both is a bad idea too already. I would argue both is ok, if properly documented. Maybe also adding explicit methods for either variant.

My argument for handling the overflow being okay is that it happens at such large values (an interval of many many days), that the precision loss is pretty irrelevant in that region.

3n-k1 commented:
> My argument for handling the overflow being okay is that it happens at such large values (an interval of many many days), that the precision loss is pretty irrelevant in that region.

that's actually a really good point. realizing that i'm definitely fine handling overflow, but i think it might be a good idea to expose methods for the variants just in case someone has a weird specific edge case

3n-k1 commented:
apologies for leaving this silent for so long, been pretty busy. i'll try and finish this pr tonight

jhass commented on src/g_lib/g_lib.cr:
```diff
@@ -1,28 +1,52 @@
-    )
+    sourcefunc = LibGLib::SourceFunc.new(trampoline.pointer, Pointer(Void).null)
+    closure = GObject::ClosureDataManager.register(trampoline.closure_data)
+    if time.milliseconds
```
This is just always truthy :)

Let's do

```cr
if time.to_i == time.total_seconds || time.total_milliseconds > UInt32::MAX
  timeout_add_seconds
elsif time.total_milliseconds > 0
  timeout_add
else
  raise ArgumentError.new("Timeout must be at least 1 millisecond")
end
```

And I agree we should also add `milliseconds_timeout` and `seconds_timeout` that always choose `timeout_add` and `timeout_add_seconds` respectively :)

3n-k1 replied:
huh, i didn't realize that 0 was truthy. i think your idea looks a lot better, too

jhass replied:
Oh and as said we should throughoutly document this in a doc comment for this method :)

3n-k1 commented:
i opted for `timeout_seconds` and `timeout_milliseconds`, so that those two methods are grouped with `timeout` in documentation or an editor's autocomplete

jhass commented on src/g_lib/g_lib.cr:
```diff
@@ -1,28 +1,63 @@
+    if time.to_i == time.total_seconds || time.total_milliseconds > UInt32::Max
+      self.timeout_seconds(time.total_seconds, priority, &block)
+    elsif time.total_milliseconds > 0
+      self.timeout_milliseconds(time.total_milliseconds, priority, &block)
```
Huh, Crystal autocasts here? I wouldn't have thought :D

3n-k1 replied:
i always think that the `Time::Span.total_*` methods return integers, no idea why :p

jhass commented on src/g_lib/g_lib.cr:
```diff
@@ -1,28 +1,63 @@
+    elsif time.total_milliseconds > 0
+      self.timeout_milliseconds(time.total_milliseconds, priority, &block)
+    else
+      raise ArgumentError.new("Timeout must be at least 1 millisecond")
```
Let's move this to `timeout_milliseconds` (`raise ... unless interval > 0`)  and add a similar check to `timeout_seconds actually :)

jhass reviewed (changes requested):
The above, doc comments and then this is perfect 🙇

3n-k1 commented:
added some doc comments, and fixed some typos i saw. sorry if there are formatting issues; vim apparently hates crystal's indenting idioms but i think i caught them all

jhass commented on src/g_lib/g_lib.cr:
```diff
@@ -1,28 +1,69 @@
   end
+
+  # Calls *&block* every *interval* until *&block* returns false.
+  # If milliseconds are not specified, `#timeout_seconds` is called, else `#timeout_milliseconds` is called
```
```suggestion
  # If milliseconds are not specified or the interval is bigger than `UInt32::Max` milliseconds, `#timeout_seconds` is called, else `#timeout_milliseconds` is called.
```

jhass commented:
🙏

3n-k1 commented:
hooray! also i'm really really sorry that i just noticed this, but [this line](https://github.com/jhass/crystal-gobject/pull/79/files#diff-7a7c90a0a5c47b127629f6d7ee7d8de135e8c81a6434aba84bb819e88f46fb4aR61) should say `UInt32::MAX` and not `UInt32::Max` :grimacing:

jhass commented:
Duh. Maybe we should find some way to test this :D Maybe adding something `samples` to make sure it compiles :) Wanna follow up with that?
